File: src/api/cpp/GravityRequestor.cpp
/** (C) Copyright 2013, Applied Physical Sciences Corp., A General Dynamics Company
 **
 ** Gravity is free software; you can redistribute it and/or modify
 ** it under the terms of the GNU Lesser General Public License as published by
 ** the Free Software Foundation; either version 3 of the License, or
 ** (at your option) any later version.
 **
 ** This program is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ** GNU Lesser General Public License for more details.
 **
 ** You should have received a copy of the GNU Lesser General Public
 ** License along with this program;
 ** If not, see <http://www.gnu.org/licenses/>.
 **
 */

/*
 * GravityRequestor.cpp
 *
 *  Created on: Aug 20, 2012
 *      Author: Chris Brundick
 */

#include "GravityRequestor.h"
#include "GravityLogger.h"
#include "spdlog/spdlog.h"

namespace gravity {

using namespace std;

GRAVITY_API GravityRequestor::~GravityRequestor() {}

GRAVITY_API void GravityRequestor::requestTimeout(std::string serviceID, std::string requestID)
{
    spdlog::get("GravityLogger")->warn("Request timed out: service id = {}, request id = {}", serviceID, requestID);
}

}

File: src/api/rust/src/gravity_requestor.rs

use crate::gravity_data_product::GravityDataProduct;

/// Interface specification for a trait object that will function as the "client" side of a request-response interaction.
pub trait GravityRequestor {
    /// Called when a response to a request is received through the Gravity infrastructure.
    fn request_filled(&mut self, service_id: &str, request_id: &str, response: &GravityDataProduct);

    /// Called when the response to a request has timed out.
    fn request_timeout(&mut self, service_id: &str, request_id: &str);
}


pub(crate) struct RequestorWrap {
    pub(crate) requestor: Box<dyn GravityRequestor>,
}